File: exercises/java/002_adapted_response_pattern/src/main/java/com/akka/training/CoffeeMachineActor.java
package com.akka.training;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;

public class CoffeeMachineActor {
  public static Behavior<CoffeeMachineCommand> create() {
    return Behaviors.setup(CoffeeMachineActor::idle);
  }

  private static Behavior<CoffeeMachineCommand> idle(
      final ActorContext<CoffeeMachineCommand> context) {

    context.getLog().info("CoffeeMachine: IDLE");

    return Behaviors.receive(CoffeeMachineCommand.class)
        .onMessage(BrewCoffee.class, command -> brewing(context, command))
        /* Can't pick up coffee until coffee is ready, stay in same behavior (equivalent to ignore the message) */
        .onMessage(PickupCoffee.class, command -> Behaviors.same())
        .build();
  }

  private static Behavior<CoffeeMachineCommand> brewing(
      final ActorContext<CoffeeMachineCommand> context, final BrewCoffee command) {

    context.getLog().info("CoffeeMachine: Brewing 1 {}", command.coffee());
    // Warn: Don't Thread.sleep in Akka actors, it utilizes a thread from the Thread pool.
    // We will see how to replace Thread.sleep by proper non-blocking scheduling in a further
    // exercise.
    try {
      Thread.sleep(10000);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }

    return coffeeReady(context, command.coffee());
  }

  private static Behavior<CoffeeMachineCommand> coffeeReady(
      final ActorContext<CoffeeMachineCommand> context, final Coffee coffee) {

    context.getLog().info("CoffeeMachine: Coffee {} is ready", coffee);
    return Behaviors.receive(CoffeeMachineCommand.class)
        /* Can't brew a new coffee until the ready one is picked-up, stay in same behavior (equivalent to ignore the message) */
        .onMessage(BrewCoffee.class, command -> Behaviors.same())
        .onMessage(PickupCoffee.class, command -> idle(context))
        .build();
  }

  // <- Protocol definition
  public sealed interface CoffeeMachineCommand permits BrewCoffee, CoffeeReady, PickupCoffee {}

  public record BrewCoffee(Coffee coffee, ActorRef<CoffeeReady> replyTo)
      implements CoffeeMachineCommand {}

  public record CoffeeReady(Coffee coffee) implements CoffeeMachineCommand {}

  /* Represents the Barista picking up the coffee and resetting the coffee machine, so that it's ready for the next coffee */
  public record PickupCoffee() implements CoffeeMachineCommand {}
  // Protocol definition ->
}

File: exercises/scala/001_request-response-pattern/src/main/scala/dev/fResult/akka/coffeeMachine/CoffeeMachineActor.scala
package dev.fResult.akka.coffeeMachine

import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
import dev.fResult.akka.Coffee
import dev.fResult.akka.barista.BaristaCommand.CoffeeReady
import dev.fResult.akka.coffeeMachine.CoffeeMachineCommand.{BrewCoffee, PickupCoffee}

import scala.util.{Failure, Try}

object CoffeeMachineActor:
  def apply(): Behavior[CoffeeMachineCommand] = idle()

  private def idle(): Behavior[CoffeeMachineCommand] = Behaviors.setup { context =>
    context.log.info("CoffeeMachine: IDLE")

    Behaviors.receiveMessage {
      case command@BrewCoffee(_, _) => onBrewCoffee(context, command)
      case PickupCoffee => Behaviors.same
    }
  }

  private def onBrewCoffee(context: ActorContext[CoffeeMachineCommand], command: BrewCoffee): Behavior[CoffeeMachineCommand] = {
    val coffee = command.coffee
    context.log.info(s"CoffeeMachine: Brewing 1 $coffee")

    Try(Thread.sleep(3000)) match {
      case Failure(ex: InterruptedException) => ex.printStackTrace()
      case _ =>
    }

    command.replyTo ! CoffeeReady(coffee)

    coffeeReady(context, command.coffee)
  }

  private def coffeeReady(context: ActorContext[CoffeeMachineCommand], coffee: Coffee): Behavior[CoffeeMachineCommand] = {
    context.log.info("CoffeeMachine: Coffee {} is ready", coffee)

    Behaviors.receiveMessage {
      case command@BrewCoffee(_, _) => onBrewCoffee(context, command)
      case PickupCoffee => onPickupCoffee(context, coffee)
    }
  }

  private def onPickupCoffee(context: ActorContext[CoffeeMachineCommand], coffee: Coffee): Behavior[CoffeeMachineCommand] = {
    context.log.info(s"CoffeeMachine: Picking up $coffee")
    idle()
  }
end CoffeeMachineActor
